src/hotspot/cpu/aarch64/codeBuffer_aarch64.cpp

```diff
@@ -42,46 +42,44 @@ void CodeBuffer::share_trampoline_for(address dest, int caller_offset) {
   _finalize_stubs = true;
 }
 
+#define __ masm.
+
 static bool emit_shared_trampolines(CodeBuffer* cb, CodeBuffer::SharedTrampolineRequests* requests) {
   if (requests == nullptr) {
     return true;
   }
 
   MacroAssembler masm(cb);
 
-  bool p_succeeded = true;
   auto emit = [&](address dest, const CodeBuffer::Offsets &offsets) {
-    masm.set_code_section(cb->stubs());
-    if (!is_aligned(masm.offset(), wordSize)) {
-      if (cb->stubs()->maybe_expand_to_ensure_remaining(NativeInstruction::instruction_size) && cb->blob() == NULL) {
-        ciEnv::current()->record_failure("CodeCache is full");
-        p_succeeded = false;
-        return p_succeeded;
-      }
-      masm.align(wordSize);
-    }
-
+    assert(cb->stubs()->remaining() >= MacroAssembler::trampoline_stub_size(), "pre-allocated trampolines");
     LinkedListIterator<int> it(offsets.head());
     int offset = *it.next();
-    for (; !it.is_empty(); offset = *it.next()) {
-      masm.relocate(trampoline_stub_Relocation::spec(cb->insts()->start() + offset));
-    }
-    masm.set_code_section(cb->insts());
+    address stub = __ emit_trampoline_stub(offset, dest);
+    assert(stub, "pre-allocated trampolines");
 
-    address stub = masm.emit_trampoline_stub(offset, dest);
-    if (stub == nullptr) {
-      ciEnv::current()->record_failure("CodeCache is full");
-      p_succeeded = false;
+    address reloc_pc = cb->stubs()->end() - NativeCallTrampolineStub::instruction_size;
+    while (!it.is_empty()) {
+      offset = *it.next();
+      address caller_pc = cb->insts()->start() + offset;
+      cb->stubs()->relocate(reloc_pc, trampoline_stub_Relocation::spec(caller_pc));
     }
-
-    return p_succeeded;
+    return true;
   };
 
-  requests->iterate(emit);
+  assert(requests->number_of_entries() >= 1, "at least one");
+  const int total_requested_size = MacroAssembler::trampoline_stub_size() * requests->number_of_entries();
```

Comment:
There's a detail here that is worth mentioning. At first I was concerned that this calculation might overestimate the required size by one 32 bit word -- that happens where we are able generate the first stub without the need for any alignment padding. However, that won't cause any unnecessary failure because the rounding up will be from an odd to an even multiple of 32 bit words i.e. adding this extra half-word will not push the request over to consume an extra 64 bit word.

Reply:
Thanks for your review, Andrew - Yes, indeed a detail here. Although the real emitted code does not get affected, when allocating the requested trampolines, the pre-allocated size the code buffer is applying for could be slightly bigger than the previous real-time processing. I am not sure how useful it is to calculate the precise allocated request size, (though doable, seems we will lose the code simplicity) for in the context the current CodeBuffer is only a temporary one, which would be finally poured into the real nmethod; and also when the code buffer is expanding[1], a quite big and heuristic size could be applied for. I dumped the `requests->number_of_entries()` here, the trampoline count: seems the value is a small number, often less than `5`, sometimes `10`, if the `ReservedCodeCacheSize > branch_range` condition is true so that trampolines are generated.

I realized maybe I should rename the `trampoline_stub_size` to something like `max_trampoline_stub_size` to reflect its semantics. I was wondering if the final version still looks okay to you?

[1] https://github.com/openjdk/jdk/blob/b4cb6c8e8b4bb10d47fd4839c7abf13a552323f6/src/hotspot/share/asm/codeBuffer.cpp#L833-L845

Reply:
@zhengxiaolinX
> I realized maybe I should rename the `trampoline_stub_size` to something like `max_trampoline_stub_size`to reflect its semantics. I was wondering if the final version still looks okay to you?

There is no need for any name change but it does no harm. My comment was only to record a detail on the PR that might help anyone who looks at this patch in future.

I agree there is no reason to complicate the size estimation code given that it is unlikely to save a significant amount of space.

So, yes, this version looks good. Ship it!